capsules/src/lpm013m126.rs

```diff
@@ -0,0 +1,587 @@
+use kernel::ErrorCode;
+
+/// Monochrome frame buffer bytes.
+/// 176 × 176 bits = 3872 bytes.
+///
+/// 2 bytes for the start of each row (command header),
+/// plus 2 bytes of data transfer period at the end
+///
+/// 176 * 2 + 2 = 354 bytes.
+pub const BUFFER_SIZE: usize = 3872 + 354;
+
+/// Best Tock can do, sadly.
+/// Would be better to have it offset.
+type SubmitBuffer<'a> = &'a mut [u8];
+
+/// Arranges frame data in a buffer
+/// whose portions can be sent directly to the device.
+struct FrameBuffer<'a> {
+    data: &'a mut [u8],
+}
+
+impl<'a> FrameBuffer<'a> {
+    /// Turns a regular buffer (back) into a FrameBuffer.
+    /// If the buffer is fresh, and the display is initialized,
+    /// this *MUST* be initialized after the call to `new`.
+    fn new(frame_buffer: &'a mut [u8]) -> Self {
+        Self { data: frame_buffer }
+    }
+
+    /// Initialize header bytes for each line.
+    fn initialize(&mut self) {
+        for i in 0..176 {
+            let line = self.get_line_mut(i);
+            let bytes = CommandHeader {
+                mode: Mode::Input1Bit,
+                gate_line: i,
+            }
+            .encode();
+            line[..2].copy_from_slice(&bytes);
+        }
+    }
+
+    /// Copy pixels from the buffer. The buffer may be shorter than frame.
+    fn blit(&mut self, buffer: &[u8], frame: &WriteFrame) {
+        if frame.column != 0 {
+            // Can't be arsed to bit shift pixels…
+            panic!("Horizontal offset not supported");
+        }
+        let rows = (frame.row)..(frame.row + frame.height);
+        // There are 8 pixels in each row per byte.
+        let sources = buffer.chunks(frame.width as usize / 8);
+        for (i, source) in rows.zip(sources) {
+            let row = self.get_row_mut(i);
+            debug!("rowl {} soul {}", row.len(), source.len());
+            row[..(source.len())].copy_from_slice(source);
+        }
+    }
+
+    /// Gets an entire raw line, ready to send.
+    fn get_line_mut(&mut self, index: u16) -> &mut [u8] {
+        const CMD: usize = 2;
+        const TRANSFER_PERIOD: usize = 2;
+        let line_bytes = CMD + 176 / 8;
+        &mut self.data[(line_bytes * index as usize)..][..line_bytes + TRANSFER_PERIOD]
+    }
+
+    /// Gets pixel data.
+    fn get_row_mut(&mut self, index: u16) -> &mut [u8] {
+        let line_bytes = 176 / 8 + 2;
+        &mut self.data[(line_bytes * index as usize + 2)..][..(176 / 8)]
+    }
+
+    /// Transform into a view of raw data for submitting to the DMA driver
+    fn with_raw_rows(
+        frame_buffer: FrameBuffer<'static>,
+        _start: u16,
+        _end: u16,
+    ) -> SubmitBuffer<'static> {
+        // HILs typically can't use offsets :/
+        // Best we can do is limit length (TODO)
+        frame_buffer.data
+    }
+}
+
+/// Modes are 6-bit, network order.
+/// They use a tree-ish encoding, so only the ones in use are listed here.
+#[allow(dead_code)]
```

Comment:
Is this unused?

Reply:
The BlinkInversion variant is present for reference but unused here. The list is not exhaustive anyway (the encoding does not map so well into a list), should unused entries get removed?

Reply:
I think it's probably best to remove the unused entries